drivers/can/can_xlnx.c

```diff
@@ -0,0 +1,1329 @@
+ *
+ * @retval always 0.
+ */
+int can_xlnx_get_max_bitrate(const struct device *dev, uint32_t *max_bitrate)
+{
+	const struct can_xlnx_dev_cfg *dev_conf = dev->config;
+
+	*max_bitrate = dev_conf->max_bitrate;
+
+	return 0;
+}
+
+/**
+ * @brief Xilinx CAN controller ISR
+ *
+ * Interrupt service routine for the Xilinx CAN controller's
+ * IRQ. Handles RX OK, TX OK, bus off and error interrupts.
+ *
+ * @param dev Pointer to the CAN controller's device.
+ */
+static void can_xlnx_isr(const struct device *dev)
+{
+	uintptr_t base_addr = DEVICE_MMIO_GET(dev);
+	uint32_t isr;
+#ifdef CONFIG_CAN_AUTO_BUS_OFF_RECOVERY
+	uint32_t sr;
+#endif
+
+	/* Get the Interrupt Status Register*/
+	isr = sys_read32(base_addr + CAN_XLNX_ISR_OFFSET);
+
+	/* Process Interrupt Status flags */
+
+	/* TX-related interrupt handling */
+	if (isr & CAN_XLNX_IRQ_MESSAGE_TX) {
+		can_xlnx_handle_tx(dev);
+	}
+
+	/* RX-related interrupt handling */
+	if (isr & CAN_XLNX_IRQ_MESSAGE_RX) {
+		can_xlnx_handle_rx(dev);
+	}
+
+	/* Error-related interrupt handling */
+	if (isr & CAN_XLNX_IRQ_MESSAGE_ERROR) {
+		can_xlnx_handle_errors(dev);
+	}
+
+	if (isr & CAN_XLNX_IRQ_BUS_OFF) {
+		/*
+		 * CAN bus off state indication is always required in order to
+		 * clear the driver instance's local error counters. This happens
+		 * automatically in hardware whenever bus off state is entered.
+		 */
+		can_xlnx_state_update(dev, CAN_BUS_OFF);
+#ifdef CONFIG_CAN_AUTO_BUS_OFF_RECOVERY
+		/*
+		 * The controller supports auto-recovery. In order to determine
+		 * when auto-recovery has completed, poll SR[ESTAT] (comp. Zynq-
+		 * 7000 TRM, Appendix B.5, p. 804) for any value other than 0x2
+		 * (indicates bus off).
+		 */
+		do {
+			sr = sys_read32(base_addr + CAN_XLNX_SR_OFFSET);
+		} while (((sr >> CAN_XLNX_SR_ERROR_STATUS_OFFSET) &
+			CAN_XLNX_SR_ERROR_STATUS_MASK) == CAN_XLNX_SR_ESTAT_BUS_OFF);
+
+		can_xlnx_state_update(dev, CAN_ERROR_ACTIVE);
+#else
+		/*
+		 * Bus off auto-recovery is disabled:
+		 * -> leave the current state at 'bus off'
+		 * -> enter configuration mode which disables further interrupts
+		 * -> restoring an operational state requires explicit transition
+		 * to normal/snoop/loopback mode
+		 * -> 'bus off' to 'error active' state transition will take place
+		 * once normal/snoop/loopback mode is entered
+		 */
+		can_xlnx_config_mode(dev);
+#endif /* CONFIG_CAN_AUTO_BUS_OFF_RECOVERY */
+
+		/* Clear the bus off interrupt bit */
+		sys_write32(CAN_XLNX_IRQ_BUS_OFF, base_addr + CAN_XLNX_ICR_OFFSET);
+	}
+}
+
+/**
+ * @brief IRQ attach function
+ *
+ * Attaches an instance of this driver to the IRQ belonging to the
+ * controller instance it manages.
+ *
+ * @param dev Pointer to the CAN controller's device.
+ */
+static void can_xlnx_configure_irq(const struct device *dev)
+{
+	const struct can_xlnx_dev_cfg *dev_conf = dev->config;
+
+	/* Attach the ISR to the respective controller's IRQ */
+	dev_conf->config_func(dev);
+}
```

Comment:
Why an entire function just for calling a function pointer?

Reply:
resolved, inlined into the init function